### DISCUSSION TITLE
How to add .js files to built bundle?

### DISCUSSION THREAD

**Discussion by xpertana** [jaredpalmer/tsdx#936] [Q&A] [answered]
I bit the bullet and learned Typescript and love tsdx so far. There is however a part of my workflow I am struggling with:
I am using the protobufjs npm package CLI to generate JavaScript source files based on a predefined .proto schema.
The output of this process is a ".js" file and associated ".js.d.ts" declaration files.

These are included in my .ts source tree in many files as follows:

```
import { IStep } from "../engage_protobuf.js"

export async function breakExec(this:any,st:IStep): Promise<RC> {
  // break command code...
 }
```

The package (standalone not React based) builds fine with ```tsdx build```

However, when I go to test the build bundle in /dist  using the node REPL I get:

```
Welcome to Node.js v12.13.1.
> x = require("./engage-jscore.cjs.development")
Thrown:
Error: Cannot find module '../../engage_protobuf.js'
Require stack:
- /Users/glenns/xpertana/engage-core/engage-core/dist/engage-jscore.cjs.development.js
- <repl>
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
    at Function.Module._load (internal/modules/cjs/loader.js:690:27)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (internal/modules/cjs/helpers.js:74:18) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/glenns/xpertana/engage-core/engage-core/dist/engage-jscore.cjs.development.js',
    '<repl>'
  ]
}
```

I was expected all dependencies to be pulled into the final bundle; Is there some rollup config "tweak" needed under the covers to pull this .js file in?  

Thanks

**Comment by agilgur5** [comment, marked as answer]
It's hard to tell what the issue is without a repro or seeing the built output. 

TSDX does support JS and I personally put in several PRs for this and there are tests for it too.

> ```ts
> import { IStep } from "../engage_protobuf.js"
> ```

I suspect the issue may be that in TS one does not use file extensions, so your import maybe isn't getting recognized as result. I would remove the `.js` from your import and make sure you've set [`allowJs`](https://www.typescriptlang.org/tsconfig/#allowJs) to `true` in your `tsconfig.json`

**Reply by xpertana** [reply]
Thanks for the guidance. You were correct...among a bunch of "import" statements deep in my source tree one file had remaining "require" which did not like the .js extension. This error showed up only at run time. Phew. Onward in my Typescript journey. It has been frustrating  in some ways but sort of "addicting" in a way: it imparts confidence as you code. Thanks again.

**Reply by agilgur5** [reply]
Yea it's a bit strange it didn't pop up during build-time. It might be because you were importing an interface/type; meaning it actually imports from a `.d.ts` declaration file and those are handled a bit weirdly in Rollup (and have some bugs in upstream rpts2) since they're not actual source code and aren't actually bundled (c.f. #744 and upstream https://github.com/ezolenko/rollup-plugin-typescript2/issues/85).

But glad it worked, and yea, that's a good summary of the TS experience for most folks 🙂
